### PR TITLE
Adding Parallel Binary Search module

### PR DESCRIPTION
**Pull request by Muaath5**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes (#3624)

I did this pull request because the previous one was closed by stale (#3777 )

### CODE REVIEW

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 1]
```diff
@@ -0,0 +1,132 @@
+---
+id: parallel-binary-search
+title: 'Parallel Binary Search'
+author: Muaath Alqarni
+description: 'Answering multiple binary search queries parallelly'
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:5]: "parallelly"

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+prerequisites:
+  - binary-search
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:7]: maybe D&Q too?

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+frequency: 0
+---
+
+
+## Resources
+
+<Resources>
+	<Resource source="CF" title="Parallel Binary Seach [tutorial]" url="45578" />
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:15]: Please put 2.1 of this book in resources 🙏
https://kostka.dev/sp/spbook.pdf

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+</Resources>
+
+<FocusProblem problem="sam" />
+
+## Solution - New Roads Queries
+Let's think how to binary search a single query.
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:21]:
```suggestion
Let's think about how to binary search a single query.
```

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+It's clear that we can use DSU with linear search, but also we can do binary search with DSU in $O(N \cdot \log_2{N})$.
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:22]: Don't say "it's clear", demonstrate it, or add a link to another solution explanation for that

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+Now let's use parallel binary search:
+Sweep through the array of edges and union them in the DSU,
+whenevery you find a query its median = $i$, if the two nodes where connected, it means shrink to left, otherwise shrink to right.
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:26]:
```suggestion
whenever there is a query with median = $i$, if the two nodes where connected, it means shrink to left, otherwise shrink to right.
```
What does shrink mean? What's the dividing factor? Also you mentioned $i$, so use it in your explanation.

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+You need to do the sweep $O(\log_2{N})$ times, so you know the exact time the nodes where connected.
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:28]: use "we" instead of "you". No real reason for this change, but it's to keep modules consistent

Also what does "so you know the exact time the nodes where connected." means? What does "time" represent?

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	int cnt[N], par[N];
+	void init(int sz) {
+		ccc = sz;
+		for (int i = 0; i <= sz; i++) par[i] = i, cnt[i] = 1;
+	}
+	int root(int node) {
+		if (par[node] == node) return node;
+		return root(par[node]);
+	}
+	bool link(int x, int y) {
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:62]: you used "union" in your explanation. Be consistent

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+		if (a > b) swap(a, b);
+		edges.push_back({a, b});
+	}
+
+	for (int i = 0; i < q; i++) {
+		cin >> qrs[i].a >> qrs[i].b;
+		qrs[i].idx = i;
+		qrs[i].l = 1, qrs[i].r = m + 1;
+	}
+	dsu cur;
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:96]: please rename it from `cur`

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+	int lg = ceil(log2(m));
+	while (lg--) {
+		cur.init(n);
+		sort(qrs, qrs + q,
+		     [](const pbs &x, const pbs &y) { return x.mid() < y.mid(); });
+		int idx = 0;
+		for (int i = 0; i < m; i++) {
+			cur.link(edges[i].first, edges[i].second);
+			while (idx < q && qrs[idx].mid() == i + 1) {
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:105]: since this is the introductory problem, some comments are required

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+				else qrs[idx].l = qrs[idx].mid() + 1;
+				idx++;
+			}
+			if (idx == q) break;
+		}
+	}
+	sort(qrs, qrs + q,
+	     [](const pbs &x, const pbs &y) { return x.idx < y.idx; });
+	for (int i = 0; i < q; i++)
+		cout << (qrs[i].l == m + 1 ? -1 : (qrs[i].a == qrs[i].b ? 0 : qrs[i].l))
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:121]: please attempt to make this readable

**Review by envyaims — CHANGES_REQUESTED** [review]

File: content/6_Advanced/Parallel_Binary_Search.mdx (hunk 1 above)
Comment (Muaath5) [reply content/6_Advanced/Parallel_Binary_Search.mdx:5]: is it wrong? how to correct?

File: content/6_Advanced/Parallel_Binary_Search.mdx (hunk 2 above)
Comment (Muaath5) [reply content/6_Advanced/Parallel_Binary_Search.mdx:7]: you can do PBS iteratively, so it's not a request I think.

File: content/6_Advanced/Parallel_Binary_Search.mdx (hunk 7 above)
Comment (Muaath5) [reply content/6_Advanced/Parallel_Binary_Search.mdx:28]: I should replace time with "amount of days", same as in the problem.

File: content/6_Advanced/Parallel_Binary_Search.mdx (hunk 1 above)
Comment (envyaims) [reply content/6_Advanced/Parallel_Binary_Search.mdx:5]: rewrite the description

File: content/6_Advanced/Parallel_Binary_Search.mdx (hunk 2 above)
Comment (envyaims) [reply content/6_Advanced/Parallel_Binary_Search.mdx:7]: the idea is similar to D&Q no? Iterative implemention doesnt mean the idea isnt fundamentally s

Comment (Muaath5) [reply content/6_Advanced/Parallel_Binary_Search.mdx:7]: it's pretty similar yeah

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 12]
```diff
@@ -0,0 +1,141 @@
+---
+id: parallel-binary-search
+title: 'Parallel Binary Search'
+author: Muaath Alqarni
+description: 'Offline way to do multiple binary search on answer efficiently'
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:5]:
```suggestion
description: 'Answer binary search queries offline'
```

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+<Resources>
+	<Resource source="CF" title="Parallel Binary Seach [tutorial]" url="45578" />
+	<Resource source="SPBook" title="Parallel Binary Seach" url="https://kostka.dev/sp/spbook.pdf#page=22" />
+</Resources>
+
+<FocusProblem problem="sam" />
+
+## Solution - New Roads Queries
+Let's think about how to use binary to answer single query.
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:23]:
```suggestion
```

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+We can see that connvectivity of the nodes is a monotonic function.
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:24]:
```suggestion
We can see that connectivity of the nodes is a monotonic function.
```

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+
+Now let's use parallel binary search:
+
+First, sort the queries by median.
+
+Then sweep through the array of edges and union them in the DSU,
+whenevery there is a query with median = $i$, if the two nodes where connected in DSU, it means the answer is $\leq i$, otherwise it means $\gt i$.
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:31]:
```suggestion
Whenever there is a query with median = $i$, if the two nodes where connected in DSU, it means the answer is $\leq i$, otherwise it means $\gt i$.
```

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+
+## Solution - New Roads Queries
+Let's think about how to use binary to answer single query.
+We can see that connvectivity of the nodes is a monotonic function.
+
+Now let's use parallel binary search:
+
+First, sort the queries by median.
+
+Then sweep through the array of edges and union them in the DSU,
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:30]:
```suggestion
Then sweep through the array of edges and union them in the DSU.
```

**Review by envyaims — CHANGES_REQUESTED** [review]
looks mostly fine!

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+		cnt[x] += cnt[y];
+	}
+	bool Connected(int x, int y) { return Root(x) == Root(y); }
+};
+
+// Parallel Binary Search Query
+struct PBS {
+	// l, r are the binary search variables
+	int idx, l, r;
+	int a, b;
```
Comment (envyaims) [inline content/6_Advanced/Parallel_Binary_Search.mdx:75]: rename to `u` and `v`, just standard convention

**Review by envyaims — APPROVED** [review]
mighty fine

File: content/6_Advanced/Parallel_Binary_Search.mdx (hunk 17 above)
Comment (SansPapyrus683) [reply content/6_Advanced/Parallel_Binary_Search.mdx:75]: i mean depends on the problem statement
unless this is an arbitrary struct

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+	<Resource source="CF" title="Parallel Binary Seach [tutorial]" url="45578" />
+	<Resource source="SPBook" title="Parallel Binary Seach" url="https://kostka.dev/sp/spbook.pdf#page=22" />
+</Resources>
+
+<FocusProblem problem="sam" />
+
+## Solution - New Roads Queries
+We can see that connectivity of the nodes is a monotonic function.
+
+Now let's use parallel binary search:
```
Comment (SansPapyrus683) [inline content/6_Advanced/Parallel_Binary_Search.mdx:25]: give a quick summary on what pbs is and how one would use it
bc right now this is not different from a normal problem editorial

File: content/6_Advanced/Parallel_Binary_Search.mdx (hunk 17 above)
Comment (Muaath5) [reply content/6_Advanced/Parallel_Binary_Search.mdx:75]: I used a,b same as the problem statement

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+
+<Resources>
+	<Resource source="CF" title="Parallel Binary Seach [tutorial]" url="45578" />
+	<Resource source="SPBook" title="Parallel Binary Seach" url="https://kostka.dev/sp/spbook.pdf#page=22" />
+</Resources>
+
+## Tutorial
+This technique is usually used when we can solve a single query by doing linear or binary search on monotonic function.
+
+The trick is doing it offline, sort queries by their median (initially $l$, $r$ will be the binary search values), do linear search,
```
Comment (jessechoe10) [inline content/6_Advanced/Parallel_Binary_Search.mdx:23]:
```suggestion
The trick is doing it offline, sorting queries by their median (initially $l$, $r$ will be the binary search values), linearly searching
```

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+whenever you find a query its median in same as your position in linear search, check the monotonic function, and do same as binary search (edit either $l$ or $r$).
```
Comment (jessechoe10) [inline content/6_Advanced/Parallel_Binary_Search.mdx:24]:
```suggestion
whenever you find a query its median in same as your position in linear search, checking the monotonic function, and binary searching (edit either $l$ or $r$).
```

Comment (jessechoe10) [inline content/6_Advanced/Parallel_Binary_Search.mdx:24]: If that's what you meant, then resolve those changes. If not, then please rewrite that sentence since it doesn't really make sense right now.

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+
+You will need to do the linear search $O(\log_2{N})$ times, so $l = r$.
```
Comment (jessechoe10) [inline content/6_Advanced/Parallel_Binary_Search.mdx:26]: Rewrite this... It sounds weird

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+
+Usually the complexity would be: $O((N+Q) \cdot \log_2{(N+Q)}^2)$
```
Comment (jessechoe10) [inline content/6_Advanced/Parallel_Binary_Search.mdx:28]: This is an incomplete thought. Usually it's that time complexity but you should also state the time complexity of the parallel binary search following a ", but"

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+
+<FocusProblem problem="sam" />
+
+## Solution - New Roads Queries
+We can see that connectivity of the nodes is a monotonic function.
```
Comment (jessechoe10) [inline content/6_Advanced/Parallel_Binary_Search.mdx:33]: How

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+
+Now let's use parallel binary search:
```
Comment (jessechoe10) [inline content/6_Advanced/Parallel_Binary_Search.mdx:35]: Combine these sentences in some way

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+
+When sweeping through the array of edges, union them in the DSU.
+
+Whenever there is a query with median = $i$, if the two nodes where connected in DSU, it means the answer is $\leq i$ ($r = i$), otherwise it means $\gt i$ ($l = i+1$).
```
Comment (jessechoe10) [inline content/6_Advanced/Parallel_Binary_Search.mdx:39]: Split this into two sentences since it's too wordy and grammatically incorrect

File: content/6_Advanced/Parallel_Binary_Search.mdx [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+The trick is doing it offline, sort queries by their median (initially $l$, $r$ will be the binary search values), do linear search,
+whenever you find a query its median in same as your position in linear search, check the monotonic function, and do same as binary search (edit either $l$ or $r$).
+
+You will need to do the linear search $O(\log_2{N})$ times, so $l = r$.
+
+Usually the complexity would be: $O((N+Q) \cdot \log_2{(N+Q)}^2)$
+
+<FocusProblem problem="sam" />
+
+## Solution - New Roads Queries
```
Comment (jessechoe10) [inline content/6_Advanced/Parallel_Binary_Search.mdx:32]: Definitely expand on this solution

**Review by jessechoe10 — CHANGES_REQUESTED** [review]
Please rewrite everything since most of it doesn't really make sense to me right now (a lot of grammatical issues), and the explanations are incomplete and can be more elaborated, especially for someone who doesn't know the topic.
